semantic-tags/src/Tags/Tagging/Precise.hs

```diff
@@ -0,0 +1,73 @@
+{-# LANGUAGE AllowAmbiguousTypes, ConstraintKinds, FlexibleContexts, FlexibleInstances, MultiParamTypeClasses, RankNTypes, ScopedTypeVariables, TypeApplications, TypeOperators #-}
+module Tags.Tagging.Precise
+( Tags
+, ToTags(..)
+, yield
+, runTagging
+, GFoldable1(..)
+) where
+
+import Control.Effect.Pure
+import Control.Effect.Reader
+import Control.Effect.Writer
+import Data.Monoid (Endo(..))
+import GHC.Generics
+import Prelude hiding (span)
+import Source.Loc (Loc)
+import Source.Span
+import Source.Source
+import Tags.Tag
+
+type Tags = Endo [Tag]
```

Comment:
A stream on the cheap. (We don’t need an effectful stream at present, so this is by far the simplest solution that doesn’t suffer O(n²) slowdowns on append.)

Reply:
I think it might be more indicative of purpose to use `Data.DList` here rather than `Endo Tag`.

Reply:
To be honest, I don’t want to pull in a dependency for something which we use exclusively via its `Monoid` instance—which is the bit that `Endo` provides for us. I think we’re providing a useful enough level of abstraction here to not worry about it.

Reply:
I think we already depend transitively on `dlist`. But it’s not a huge deal. Can you put a comment here that `Endo` is so we get fast append?